src/contracts/adapters/wormhole/WormholeAdapter.sol

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import {SafeCast} from 'openzeppelin-contracts/contracts/utils/math/SafeCast.sol';
+import {IWormholeReceiver} from './interfaces/IWormholeReceiver.sol';
+import {IWormholeRelayer} from './interfaces/IWormholeRelayer.sol';
+import {IWormholeAdapter} from './IWormholeAdapter.sol';
+import {BaseAdapter, IBaseAdapter} from '../BaseAdapter.sol';
+import {Errors} from '../../libs/Errors.sol';
+import {ChainIds} from '../../libs/ChainIds.sol';
+
+/**
+ * @title WormholeAdapter
+ * @author BGD Labs
+ * @notice Wormhole bridge adapter. Used to send and receive messages cross chain
+ * @dev it uses the eth balance of CrossChainController contract to pay for message bridging as the method to bridge
+        is called via delegate call
+ */
+contract WormholeAdapter is BaseAdapter, IWormholeAdapter, IWormholeReceiver {
+  /// @inheritdoc IWormholeAdapter
+  address public immutable WORMHOLE_RELAYER;
+
+  /// @inheritdoc IWormholeAdapter
+  address public immutable REFUND_ADDRESS;
+
+  /**
+   * @notice only calls from the set relayer are accepted.
+   */
+  modifier onlyRelayer() {
+    require(msg.sender == WORMHOLE_RELAYER, Errors.CALLER_NOT_WORMHOLE_RELAYER);
+    _;
+  }
+
+  /**
+   * @param crossChainController address of the cross chain controller that will use this bridge adapter
+   * @param wormholeRelayer wormhole entry point address
+   * @param refundAddress address that will receive left over gas
+   * @param trustedRemotes list of remote configurations to set as trusted
+   */
+  constructor(
+    address crossChainController,
+    address wormholeRelayer,
+    address refundAddress,
+    TrustedRemotesConfig[] memory trustedRemotes
+  ) BaseAdapter(crossChainController, trustedRemotes) {
+    require(wormholeRelayer != address(0), Errors.WORMHOLE_RELAYER_CANT_BE_ADDRESS_0);
+    WORMHOLE_RELAYER = wormholeRelayer;
+    REFUND_ADDRESS = refundAddress;
+  }
+
+  /// @inheritdoc IBaseAdapter
+  function forwardMessage(
+    address receiver,
+    uint256 destinationGasLimit,
+    uint256 destinationChainId,
+    bytes calldata message
+  ) external returns (address, uint256) {
+    uint16 nativeChainId = SafeCast.toUint16(infraToNativeChainId(destinationChainId));
+    require(nativeChainId != uint16(0), Errors.DESTINATION_CHAIN_ID_NOT_SUPPORTED);
+    require(receiver != address(0), Errors.RECEIVER_NOT_SET);
+
+    uint64 sequence = IWormholeRelayer(WORMHOLE_RELAYER).sendPayloadToEvm{
+      value: _getGasCost(destinationGasLimit, nativeChainId)
+    }(
+      nativeChainId,
+      receiver,
+      message,
+      0, // no receiver value needed
+      destinationGasLimit,
+      nativeChainId,
+      REFUND_ADDRESS
+    );
+
+    return (address(WORMHOLE_RELAYER), uint256(sequence));
+  }
+
+  /// @inheritdoc IWormholeReceiver
+  function receiveWormholeMessages(
+    bytes calldata payload, // TODO: does it affect?? if in interface its as memory??
```

Comment:
I would keep memory as well then

Reply:
problem is that the subsequent call expects calldata. From the tests on testnets it works, so should we leave it as is?

Reply:
ok